composite_broker/src/broker.rs

```diff
@@ -0,0 +1,298 @@
+                },
+                mqtt_sn::TopicNameOrId::Name(name) => {
+                    // generate a random number that is not a key in the list already
+                    // println!("Topic is : {}", name.to_string());
+                    let mut rng = rand::thread_rng();
+                    let mut n2: u16 = rng.gen_range(0..999);
+                    while self.concrete_subscriptions_list.contains_key(&n2) {
+                        n2 = rng.gen()
+                    }
+                    // add it to the pairs list
+                    self.topicname_id_pairs.insert(name.clone().to_string(), n2.clone());
+                    topic = n2.clone()
+                }
+            };
+            match self.concrete_subscriptions_list.get_mut(&topic) {
+                Some(entry) => {
+                    entry.push(Client { client_id: cli.client_id.clone(), address: addr_port[0].to_string(), port: addr_port[1].to_string()});
+                },
+                None => {
+                    self.concrete_subscriptions_list.insert(topic, vec![Client { client_id: cli.client_id.clone(), address: addr_port[0].to_string(), port: addr_port[1].to_string() }]);
+                }
+            }
+            Message::SubAck(SubAck { 
+                flags: Flags::default(), 
+                msg_id: self.num_packets, 
+                topic_id: topic,
+                code: ReturnCode::Accepted })
+
+        } // end subscribe
+
+        // return the client address (String) and the publish ack packet
+        pub fn accept_pub(&mut self, addr: String, packet: Publish) -> (Message, Message, Vec<String>) {
+            self.num_packets +=1;
+            let addr_port: Vec<&str> = addr.split(":").collect();
+            let pub_p = Message::Publish(Publish { 
+                ..packet.clone()
+            });
+
+            if !self.find_client_by_address(addr_port[0], addr_port[1]) {
+                println!("Address not in broker");
+                let ack = Message::PubAck(PubAck { 
+                    topic_id: packet.topic_id, 
+                    msg_id: self.num_packets, 
+                    code: ReturnCode::Rejected(RejectedReason::NotSupported) 
+                });
+                return (ack, pub_p, Vec::new())
+            }
+
+            // create return variables
+            let mut ret_clients : Vec<String> = Vec::new();
+            let mut found = false;
+            let topic = &packet.topic_id;           
+            let top = self.concrete_subscriptions_list.get(topic);
+
+            // find the topic in the subscriptions list
+            match top {
+                Some(list) => {
+                    found = true;
+                    println!("\tFound the topic");
+                    // add the addresses for the clients that are subscribed to the topic to the return vector
+                    for cli in list {
+                        ret_clients.push(cli.address.clone()+":"+&cli.port.clone());
```

Comment:
Anotehr example where all of this is a little more complicated (and slower) when using the strings.